File: Hard/Python/719. Find K-th Smallest Pair Distance.py
class Solution(object):
    def smallestDistancePair(self, nums, k):
        # Sort the list, then initialize a minimum and maximum variable
        nums.sort()
        minimum = 0
        maximum = nums[len(nums) - 1] - nums[0]

        # Binary search algorithm
        while minimum < maximum:
            mid = (minimum + maximum) / 2
            count = 0
            i = 0

            # Find pairs with distance less than or equal to middle
            for j in range(len(nums)):
                # Run through this segment of code once, then break
                while True:
                    diff = nums[j] - nums[i]
                    if mid < diff:
                        i += 1
                    else:
                        break
                count += j - i
        
            if k <= count:
                maximum = mid
            else:
                minimum = mid + 1
        
        # Return the smallest distance found
        return minimum

File: Hard/Java/719. Find K-th Smallest Pair Distance.java
class Solution {
    public int smallestDistancePair(int[] nums, int k) {
        // Sort the array, then initialize a min and max variable
        Arrays.sort(nums);
        int min = 0;
        int max = nums[nums.length - 1] - nums[0];

        // Binary search algorithm
        while(min < max){
            int mid = (min + max) / 2;
            int count = 0;
            int i = 0;

            // Find pairs with distance less than or equal to middle
            for(int j = 0; j < nums.length; j++){
                // Run through this segment of the code once, then break
                while(true){
                    int diff = nums[j] - nums[i];
                    if(mid < diff)
                        i++;
                    else
                        break;
                }
                count += j - i;
            }
            
            if(k <= count)
                max = mid;
            else
                min = mid + 1;
        }

        // Return the smallest distance found
        return min;
    }
}
